### PR TITLE
Microsoft.Data.Sqlite: Improve SqliteDataReader.GetFieldType() when NULL

### LINKED ISSUE #13831: Microsoft.Data.Sqlite: Improve SqliteDataReader.GetFieldType() when NULL
`SqliteDataReader.GetFieldType(0)` returns `int` when the value is `NULL`. We should improve this.

Thoughts:
* This should probably return `DBNull`, but will that break other assumptions in ADO.NET?
* If we simply don't have enough information to determine the type, should we...
  * ...throw?
  * ...use a highly-compatible type (like `string`)?
  * ...use the type affinity of its column?

### LINKED ISSUE #13839: Microsoft.Data.Sqlite: Improve SqliteDataReader.GetSchemaTable() DataType when no rows
I'm using a custom DataAdapter i wrote to fill the void for the lack of one in this library, that is as follows:
```
using System;
using System.Data;
using System.Data.Common;
using Microsoft.Data.Sqlite;

namespace Ultz.BeagleFramework.SQLite
{
    public delegate void SqliteRowUpdatedEventHandler(Object sender, SqliteRowUpdatedEventArgs e);
    public delegate void SqliteRowUpdatingEventHandler(Object sender, SqliteRowUpdatingEventArgs e);
    public sealed class SqliteDataAdapter : DbDataAdapter
    {
        public event SqliteRowUpdatedEventHandler RowUpdated;

        public event SqliteRowUpdatingEventHandler RowUpdating;

        public SqliteDataAdapter() { }

        public SqliteDataAdapter(SqliteCommand selectCommand)
        {
            SelectCommand = selectCommand;
        }

        public SqliteDataAdapter(string selectCommandText, SqliteConnection selectConnection)
            : this(new SqliteCommand(selectCommandText, selectConnection)) { }

        public SqliteDataAdapter(string selectCommandText, string selectConnectionString)
            : this(selectCommandText, new SqliteConnection(selectConnectionString)) { }

        protected override RowUpdatedEventArgs CreateRowUpdatedEvent( DataRow dataRow,  IDbCommand command,
                                                                     System.Data.StatementType statementType,
                                                                      DataTableMapping tableMapping)
        {
            return new SqliteRowUpdatedEventArgs(dataRow, command, statementType, tableMapping);
        }

        protected override RowUpdatingEventArgs CreateRowUpdatingEvent( DataRow dataRow,  IDbCommand command,
                                                                       System.Data.StatementType statementType,
                                                                        DataTableMapping tableMapping)
        {
            return new SqliteRowUpdatingEventArgs(dataRow, command, statementType, tableMapping);
        }

        protected override void OnRowUpdated( RowUpdatedEventArgs value)
        {
            //base.OnRowUpdated(value);
            if (RowUpdated != null && value is SqliteRowUpdatedEventArgs)
                RowUpdated(this, (SqliteRowUpdatedEventArgs)value);
        }

        protected override void OnRowUpdating( RowUpdatingEventArgs value)
        {
            if (RowUpdating != null && value is SqliteRowUpdatingEventArgs)
                RowUpdating(this, (SqliteRowUpdatingEventArgs)value);
        }

        public new SqliteCommand DeleteCommand
        {
            get { return (SqliteCommand)base.DeleteCommand; }
            set { base.DeleteCommand = value; }
        }

        public new SqliteCommand SelectCommand
        {
            get { return (SqliteCommand)base.SelectCommand; }
            set { base.SelectCommand = value; }
        }

        public new SqliteCommand UpdateCommand
        {
            get { return (SqliteCommand)base.UpdateCommand; }
            set { base.UpdateCommand = value; }
        }

        public new SqliteCommand InsertCommand
        {
            get { return (SqliteCommand)base.InsertCommand; }
            set { base.InsertCommand = value; }
        }
    }

#pragma warning disable 1591

    public class SqliteRowUpdatingEventArgs : RowUpdatingEventArgs
    {
        public SqliteRowUpdatingEventArgs(DataRow dataRow, IDbCommand command, System.Data.StatementType statementType,
                                          DataTableMapping tableMapping)
            : base(dataRow, command, statementType, tableMapping) { }
    }

    public class SqliteRowUpdatedEventArgs : RowUpdatedEventArgs
    {
        public SqliteRowUpdatedEventArgs(DataRow dataRow, IDbCommand command, System.Data.StatementType statementType,
                                         DataTableMapping tableMapping)
            : base(dataRow, command, statementType, tableMapping) { }
    }

#pragma warning restore 1591
}
```

After creating a table, I then `SELECT * FROM` the table, and use `SqliteDataAdapter.Fill()`. However, all of the types are set to Int32. This also happens when using `DataTable.Load()`.

### PR DESCRIPTION
**Pull request by AlexanderTaeschner**
In case a column value is NULL, SqliteDataReader.GetFieldType() now tries to return the field type based on the column affinity. 
SQLite does not provide the affinity via its API, therefore the corresponding C function sqlite3AffinityType was ported to C#. The affinity NUMERIC has no corresponding data type, so here the data type TEXT is returned (as suggested as default in #13831). 
A special case is a column where no type is specified, here the explicit rule was used, that such a column is of affinity BLOB (see [Determination Of Column Affinity](https://www.sqlite.org/datatype3.html#determination_of_column_affinity)).
The corresponding changes were also introduced for the SqliteDataReader.GetSchemaTable() DataType column.

Fixes #13831 and fixes #13839

### CODE REVIEW

**Comment by seesharper** [comment]
Any updates on this. Will this make it into the next release?

**Comment by bricelam** [comment]
Haven’t had a chance to review it yet. It will make it into 3.0.0, but not the next preview.

Also, I’m still not sure if we want TEXT or BLOB to be the ultimate fallback.

File: src/Microsoft.Data.Sqlite.Core/SqliteDataRecord.cs [hunk 1]
```diff
@@ -288,5 +299,69 @@ private byte[] GetCachedBlob(int ordinal)
...
             return blob;
         }
+
+        private int GetColumnAffinity(int ordinal)
+        {
+            var columnType = raw.sqlite3_column_decltype(_stmt, ordinal);
+            return Sqlite3AffinityType(columnType);
+        }
+
+        internal static int Sqlite3AffinityType(string dataTypeName)
```
Comment (bricelam) [inline src/Microsoft.Data.Sqlite.Core/SqliteDataRecord.cs]: I'd rather copy the code from EF Core...

https://github.com/aspnet/EntityFrameworkCore/blob/94c06a21bae6aeaeb2734b6c01a0374e47c9a26f/src/EFCore.Sqlite.Core/Storage/Internal/SqliteTypeMappingSource.cs#L125-L137

File: src/Microsoft.Data.Sqlite.Core/SqliteDataReader.cs [hunk 2]
```diff
@@ -553,18 +554,8 @@ public override DataTable GetSchemaTable()
...
 
-                        command.Parameters.Clear();
-                        var columnType = "typeof(\"" + columnName.Replace("\"", "\"\"") + "\")";
-                        command.CommandText = new StringBuilder()
-                            .AppendLine($"SELECT {columnType}")
-                            .AppendLine($"FROM \"{tableName}\"")
-                            .AppendLine($"WHERE {columnType} != 'null'")
-                            .AppendLine($"GROUP BY {columnType}")
-                            .AppendLine("ORDER BY count() DESC")
-                            .AppendLine("LIMIT 1;").ToString();
```
Comment (bricelam) [inline src/Microsoft.Data.Sqlite.Core/SqliteDataReader.cs]: I still want to do this since it's more reliable than the type affinity.

File: src/Microsoft.Data.Sqlite.Core/SqliteDataRecord.cs [hunk 3]
```diff
@@ -288,5 +299,69 @@ private byte[] GetCachedBlob(int ordinal)
 
             return blob;
         }
+
+        private int GetColumnAffinity(int ordinal)
+        {
+            var columnType = raw.sqlite3_column_decltype(_stmt, ordinal);
```
Comment (bricelam) [inline src/Microsoft.Data.Sqlite.Core/SqliteDataRecord.cs]: I think there's already a method on SqliteDataReader that does this. (If there is, just call it.)

File: src/Microsoft.Data.Sqlite.Core/SqliteDataRecord.cs [hunk 4]
```diff
@@ -115,7 +116,7 @@ public virtual string GetDataTypeName(int ordinal)
                     return "BLOB";
 
                 case raw.SQLITE_NULL:
-                    return "INTEGER";
+                    return "BLOB"; // since no type is specified the column has affinity BLOB
```
Comment (bricelam) [inline src/Microsoft.Data.Sqlite.Core/SqliteDataRecord.cs]: I'm going to send out a separate PR to change this. There are a few other places (including in EF) that we need to change.

File: src/Microsoft.Data.Sqlite.Core/SqliteDataReader.cs (hunk 2 above)
Comment (AlexanderTaeschner) [reply src/Microsoft.Data.Sqlite.Core/SqliteDataReader.cs]: But removing this code would solve #13839.

File: src/Microsoft.Data.Sqlite.Core/SqliteDataRecord.cs (hunk 1 above)
Comment (AlexanderTaeschner) [reply src/Microsoft.Data.Sqlite.Core/SqliteDataRecord.cs]: Done.

File: src/Microsoft.Data.Sqlite.Core/SqliteDataRecord.cs (hunk 3 above)
Comment (AlexanderTaeschner) [reply src/Microsoft.Data.Sqlite.Core/SqliteDataRecord.cs]: Done.

File: src/Microsoft.Data.Sqlite.Core/SqliteDataReader.cs (hunk 2 above)
Comment (bricelam) [reply src/Microsoft.Data.Sqlite.Core/SqliteDataReader.cs]: The algorithm I want to use:

1. Get the most used type for the column
2. If none (or null), get the type affinity
3. If none, use BLOB

File: src/Microsoft.Data.Sqlite.Core/SqliteDataRecord.cs [hunk 5]
```diff
@@ -140,6 +142,16 @@ public virtual string GetDataTypeName(int ordinal)
         public virtual Type GetFieldType(int ordinal)
         {
             var sqliteType = GetSqliteType(ordinal);
+            if (sqliteType == SQLITE_NULL)
+            {
+                sqliteType = GetColumnAffinity(ordinal);
```
Comment (bricelam) [inline src/Microsoft.Data.Sqlite.Core/SqliteDataRecord.cs]: I wonder if we should cache a previous non-null value and use it--it's probably more accurate than the type affinity.

Comment (AlexanderTaeschner) [reply src/Microsoft.Data.Sqlite.Core/SqliteDataRecord.cs]: Added caching.

**Comment by bricelam** [comment]
(Rebased on top of master)
